pkg/controller/common/controller_context.go

```diff
@@ -97,8 +102,24 @@ func CreateControllerContext(cb *clients.Builder, stop <-chan struct{}, targetNa
 		APIExtInformerFactory:                               apiExtSharedInformer,
 		ConfigInformerFactory:                               configSharedInformer,
 		OperatorInformerFactory:                             operatorSharedInformer,
+		EtcdInformer:                                        etcdInformer,
 		Stop:                                                stop,
 		InformersStarted:                                    make(chan struct{}),
 		ResyncPeriod:                                        resyncPeriod(),
 	}
 }
+
+func getEtcdInformer(operatorClient operatorclientset.Interface, operatorSharedInformer operatorinformers.SharedInformerFactory) operatorv1.EtcdInformer {
+	operatorGroups, err := operatorClient.Discovery().ServerResourcesForGroupVersion("operator.openshift.io/v1")
```

Comment:
I don't understand this function. It seems like it would be better to have the controller wait for the `etcd.operator` resource to sync similar to how it does the other CRDs it needs (like proxy and infra config ones)

Reply:
Currently, cluster-etcd-operator is not part of the release payload, so the `Etcd` CRD kind is not created by the CEO. But the library-go already has the types defined in go for `Etcd` kind [here](https://github.com/openshift/machine-config-operator/blob/8fa45661c50047de097db3ed7592e48910bfe401/vendor/github.com/openshift/client-go/operator/informers/externalversions/operator/v1/interface.go#L18). 

Because the CRD for the kind is not created, the informer blocks on `HasSynced` function as it is not able to list the type. So this function checks through the Discovery API if the kind is registered. If not, it will set the informer to `nil` so we don't block on our sync loops.

Hope that answers why we cannot wait on the Etcd CRD like proxy and infra config. Open to other ideas on how this could work...

Reply:
oh and also the assumption is the eventually when cluster-etcd-operator will be part of the release payload, this code will be executed after `Etcd` CR is created by CVO.

Reply:
Ah, yes. Thank you. 

Should we add the `Etcd` CRD to the payload here in the MCO repo? We can label it `create-only: "true"` so that the CEO (or cluster-config-operator?) can take over updating it later.

Perhaps we should also include the `Etcd/cluster` resource with `managementState: "managed"` in here as well?

Reply:
Interesting idea. I think we are trying to bring in changes that are absolutely required in MCO so that we can have the independent iteration cycle. 

So if this temporary solution works, I would rather have it this way, unless you feel strongly about it.

Reply:
I'd prefer creating the CRD over adding conditional logic in the MCO. My concerns are like, you'd need to know to restart the MCO after creating the etcd CRD for it to start syncing correctly, or legitimate issues (e.g. rbac is wrong, transient api errors) won't be surfaced when the MCO should be syncing the etcd correctly.

Are you planning on creating the etcd CRD in the cluster-config-operator?

Reply:
@ericavonb Thanks for pointing out, these are legitimate concerns. As I pointed out earlier, I am relying on CVO to create the etcd CRD is cluster-etcd-operator is part of the release payload(static file [here](https://github.com/openshift/cluster-etcd-operator/blob/master/manifests/0000_12_etcd-operator_01_config.crd.yaml)).

It looks like the run level of cluster-etcd-operator is 12 and that of MCO is 80 (guessed from the file names), so I assumed that if cluster-etcd-operator is part of release payload, the etcd CRD will always be created by CVO[1] before MCO pod starts to run. Is that a wrong/dangerous assumption to make?

If the Discovery call fails out due to some transient errors, I can rework this function to throw an error and restart the MCO pod.

If you are still concerned this would not work, I can create the etcd CRD in MCO, but that might lead to a situation where the CRD is created in the system but the cluster-etcd-operator is not part of the release payload, I am not sure if it is desirable.  

1. https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level

Reply:
> It looks like the run level of cluster-etcd-operator is 12 and that of MCO is 80 (guessed from the file names), so I assumed that if cluster-etcd-operator is part of release payload, the etcd CRD will always be created by CVO[1] before MCO pod starts to run. Is that a wrong/dangerous assumption to make?

I think that should be fine to assume 👍 .

> If the Discovery call fails out due to some transient errors, I can rework this function to throw an error and restart the MCO pod.

That seems reasonable to me.